### PR TITLE
added GEMPLP

### PR DESCRIPTION
**Pull request by Jiaolong**
@hushell Could you please help me to check this patch?

### CODE REVIEW

**Comment by hushell** [comment]
@Jiaolong Good good! This one is much better than what I was thinking about. 
It's nice to get rid of Region class in the original implementation (OI). I think you have fully understand the MPLP algorithm. I can tell from update_messages(), which is almost different from the OI. You are amazing, man! 
The only thing that we should be more careful is the naming rule of variables, functions etc. I'll expand that in the code.

File: src/shogun/structure/GEMPLP.h [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+		          const float64_t int_gap_thr = 0.0002)
+			: m_max_iter(max_iter),
+			  m_obj_del_thr(obj_del_thr),
+			  m_int_gap_thr(int_gap_thr)
+		{}
+
+		/** maximum number of outer iterations*/
+		int32_t m_max_iter;
+		/** threshold of the delta objective value */
+		float64_t m_obj_del_thr;
```
Comment (hushell) [inline src/shogun/structure/GEMPLP.h]: May need to explain more about these vars. 
obj_del = last_dual_obj - dual_obj
int_gap = duality gap
Correct me if I am wrong.

File: src/shogun/structure/GEMPLP.h [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+	 */
+	void max_in_subdimension(SGNDArray<float64_t> tar_arr, SGVector<int32_t> &subset_inds, SGNDArray<float64_t> &max_res) const;
+
+	/** Find separators between clqiues
+	 *
+	 * @param clique_A clique A
+	 * @param clique_B clique B
+	 * @return index in all separators
+	 */
+	int32_t find_separator_index(SGVector<int32_t> clique_A, SGVector<int32_t> clique_B);
```
Comment (hushell) [inline src/shogun/structure/GEMPLP.h]: I guess this is analogy to sepset for clique trees. However, actually messages are passed between clusters/regions which are not formed in a hypertree but in general a hypergraph. I don't know if seperator set is a conventional in the context or not. Besides, calling it intersection sets is not a bad idea. The more we follow the OI names, the easier for people to understand.

Comment (hushell) [inline src/shogun/structure/GEMPLP.h]: I suggest to rename to find_intersection_set_index

File: src/shogun/structure/GEMPLP.h [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+	/** Convert original energies to potentials of the cluster
+	 * GEMPLP objective function is a maximazation function, we use - energy
+	 * as potential. The indices of the table factor energy also need to be
+	 * converted to the order of the nd array.
+	 *
+	 * @param factor factor which contains energy
+	 * @return potential of the cluster in MPLP
+	 */
+	SGNDArray<float64_t> convert_message(CFactor* factor);
```
Comment (hushell) [inline src/shogun/structure/GEMPLP.h]: Suggest to rename to convert_energy_to_potential

File: src/shogun/structure/GEMPLP.h [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+
+	/** Insert the new value into a vector
+	 *
+	 * @param v the target vector
+	 * @param k current value
+	 */
+	void exist_or_pushback(vector<int32_t>& v, int32_t k);
```
Comment (hushell) [inline src/shogun/structure/GEMPLP.h]: Is it better to use std:set? since you don't want to have copies. I mean: vector< set<int32_t> > m_clique_separators. Then use set:insert, rather than this function.

File: src/shogun/structure/GEMPLP.h [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+
+public:
+	/** GEMPLP parameter */
+	Parameter m_param;
+	/** all factors in the graph*/
+	CDynamicObjectArray* m_factors;
+	/** all separators */
+	vector<SGVector<int32_t> > m_all_separators;
```
Comment (hushell) [inline src/shogun/structure/GEMPLP.h]: = m_all_intersects?

File: src/shogun/structure/GEMPLP.h [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+	/** the separator indices (node indices) on each clique */
+	vector<vector<int32_t> > m_clique_separators;
```
Comment (hushell) [inline src/shogun/structure/GEMPLP.h]: It's nice to have members here since we don't have class Region. However, I suggest to rename separaters -> intersection_sets and clique -> region. Also applies to other members below.

File: src/shogun/structure/GEMPLP.h [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+	/** the indices (orders in the clique) of the separators on each clique */
+	vector<vector<SGVector<int32_t> > > m_clique_inds_separators;
+	/** store the sum of messages into separators */
+	vector<SGNDArray<float64_t> > m_msgs_into_separators;
+	/** store the messages from cluster to separator */
+	vector<vector<SGNDArray<float64_t> > > m_msgs_from_cluster;
```
Comment (hushell) [inline src/shogun/structure/GEMPLP.h]: cluster -> region

File: src/shogun/structure/GEMPLP.h [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+	/** store the original (-) energy of the factors */
+	vector<SGNDArray<float64_t> > m_theta_cluster;
```
Comment (hushell) [inline src/shogun/structure/GEMPLP.h]: -> m_region_theta
I know they call m_region_lambda, but lambda was used in the paper differently .

File: src/shogun/structure/GEMPLP.h (hunk 1 above)
Comment (Jiaolong) [inline src/shogun/structure/GEMPLP.h]: Yes,  int_gap = primal_obj_val - dual_obj_val

File: src/shogun/structure/MAPInference.cpp [hunk 9]
```diff
@@ -38,6 +39,9 @@ CMAPInference::CMAPInference(CFactorGraph* fg, EMAPInferType inference_method)
 		case GRAPH_CUT:
```
Comment (hushell) [inline src/shogun/structure/MAPInference.cpp]: Now I think it's safe to remove other inference types, only keep TreeMaxProd (for tree), GraphCut (for binary loopy graph, better with submodular energy gurantee) and GEMPLP (general loopy graph), since we don't want to implement more methods. The default one choose GEMPLP. In a short time period, we don't want to support 3rd party inference lib.  
@Jiaolong @iglesias Thoughts?

File: src/shogun/structure/GEMPLP.h (hunk 2 above)
Comment (Jiaolong) [inline src/shogun/structure/GEMPLP.h]: OK, I agree with you. I just followed DRWN. I will change the names.

File: src/shogun/structure/GEMPLP.h (hunk 3 above)
Comment (Jiaolong) [inline src/shogun/structure/GEMPLP.h]: Agree!

File: src/shogun/structure/MAPInference.cpp (hunk 9 above)
Comment (hushell) [inline src/shogun/structure/MAPInference.cpp]: BTW, since the factor graph is an input, we could check the graph type, and warn if the inference type is not a best fit.

File: src/shogun/structure/GEMPLP.h (hunk 4 above)
Comment (Jiaolong) [inline src/shogun/structure/GEMPLP.h]: Sounds very good!

File: src/shogun/structure/GEMPLP.h (hunk 5 above)
Comment (Jiaolong) [inline src/shogun/structure/GEMPLP.h]: OK

File: src/shogun/structure/GEMPLP.h (hunk 6 above)
Comment (Jiaolong) [inline src/shogun/structure/GEMPLP.h]: OK

File: src/shogun/structure/GEMPLP.h (hunk 8 above)
Comment (Jiaolong) [inline src/shogun/structure/GEMPLP.h]: Yes, so I will keep theta.

**Comment by Jiaolong** [comment]
Here I show you the results on yest dataset. The computation is slightly high.

Example 1: tree structure, max-product inference
----------------------------------------------------
SGD trained in   47.0507
sgd solver: average training loss = 0.030553
sgd solver: average testing error = 0.113155

Example 2.1: tree structure, graph-cuts inference
----------------------------------------------------
SGD trained in   42.9520
sgd solver: average training loss = 0.049546
sgd solver: average testing error = 0.125000

Example 2.2: full-connected graph, graph-cuts inference
----------------------------------------------------
SGD trained in   67.9839
sgd solver: average training loss = 0.082026
sgd solver: average testing error = 0.134894

Example 3.1: tree structure, GEMPLP inference
----------------------------------------------------
SGD trained in   76.5377
sgd solver: average training loss = 0.031104
sgd solver: average testing error = 0.113852

Example 3.2: full-connected graph, GEMPLP inference
----------------------------------------------------
SGD trained in  460.3569
sgd solver: average training loss = 0.030966
sgd solver: average testing error = 0.102425

File: src/shogun/structure/GEMPLP.cpp [hunk 10]
```diff
@@ -0,0 +1,368 @@
...
+	SGNDArray<float64_t> lam_sum(cards);
+
+	if (m_theta_cluster[id_clique].len_array == 0)
+		lam_sum.set_const(0);
+	else
+		lam_sum = m_theta_cluster[id_clique].clone();
+
+	int32_t num_separators = m_clique_separators[id_clique].size();
+	vector<SGNDArray<float64_t> > lam_minus; // substract message: \lambda_s^{-c}(x_s)
+	// \sum_{\hat{s}} \lambda_{\hat{s}}^{-c}(x_{\hat{s}}) + \theta_c(x_c)
```
Comment (hushell) [inline src/shogun/structure/GEMPLP.cpp]: Is this one the same as the eq in section 6 of [1]?

File: src/shogun/structure/GEMPLP.cpp [hunk 11]
```diff
@@ -0,0 +1,368 @@
...
+	for (int32_t s = 0; s < num_separators; s++)
+	{
+		// maximazation: \max_{x_c} \sum_{\hat{s}} \lambda_{\hat{s}}^{-c}(x_{\hat{s}}) + \theta_c(x_c)
+		SGNDArray<float64_t> lam_max(lam_minus[s].get_dimensions());
+		max_in_subdimension(lam_sum, m_clique_inds_separators[id_clique][s], lam_max);
+		int32_t id_separator = m_clique_separators[id_clique][s];
+		// weighted sum
+		lam_max *= 1.0/num_separators;
+		m_msgs_from_cluster[id_clique][s] = lam_max.clone();
+		m_msgs_from_cluster[id_clique][s] -= lam_minus[s];
```
Comment (hushell) [inline src/shogun/structure/GEMPLP.cpp]: lam_minus need to multiply (1-1/num_sep) ?

**Comment by hushell** [comment]
@Jiaolong Good job! I'll check your code out  and test it on my machine, then I probably give more comments later.

File: src/shogun/structure/MAPInference.cpp (hunk 9 above)
Comment (Jiaolong) [inline src/shogun/structure/MAPInference.cpp]: Yes, I totally agree with you. 

I am also considering to re-factor energy table and use `SGNDArray` to store energies and messages.  

Table factor operation class could also be implemented in the future because the current operation can directly done by `SGNDArray` which is not straightforward and hard to understand. Such ideas are just inspired from `DRWIN`.

What do you think?

**Comment by Jiaolong** [comment]
Here are the results from the MPLP V1.0 library. 

Example 3.1: tree structure, GEMPLP inference
----------------------------------------------------
SGD trained in   80.2897
sgd solver: average training loss = 0.040462
sgd solver: average testing error = 0.120262

Example 3.2: full-connected graph, GEMPLP inference
----------------------------------------------------
SGD trained in  213.0953
sgd solver: average training loss = 0.040462
sgd solver: average testing error = 0.120262

File: src/shogun/structure/GEMPLP.cpp (hunk 11 above)
Comment (Jiaolong) [inline src/shogun/structure/GEMPLP.cpp]: Yes, you are right. In fact, I am confused by this line of code as well. The original implementation does not  use `lam_minus *= 1.0/num_separators;`. I will check it later.

File: src/shogun/structure/MAPInference.cpp (hunk 9 above)
Comment (hushell) [inline src/shogun/structure/MAPInference.cpp]: Definitely a good idea! Some of functions in CFactorType CFactor for operating energy tables are hard to follow. You may open an issue for this. So either you or I can do it when we have time.

**Comment by hushell** [comment]
@Jiaolong Good to see these comparisons! As we expected, MPLP does a better job than GraphCuts in general graphs and similar to TreeBP in trees. Didn't expect MPLP is so much time consuming, but 460s = <8mins is acceptable? Since GraphCuts was not too bad, maybe better to set it as default.

**Comment by hushell** [comment]
"the results from the MPLP V1.0 library" is keeping their code as before? 
You may want to timing each part of the code to see why the current CGEMPLP is slower. Also check if current CGEMPLP uses more iterations?

**Comment by Jiaolong** [comment]
MPLP V1.0 library is the one that I kept all original implementation.

**Comment by iglesias** [comment]
I am sorry for the long delay, guys, but this pull request is HUGE :-)

File: tests/unit/structure/GEMPLP_unittest.cc [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+
+	EXPECT_NEAR(min_energy_expected, infer_met.get_energy(), 1E-10);
+
+	SG_UNREF(fg_random);
+	SG_UNREF(fg_test_data);
+}
+
+// Test with SOSVM
+/* This test is time consuming, uncomment it if needed
+TEST(GEMPLP, sosvm)
```
Comment (iglesias) [inline tests/unit/structure/GEMPLP_unittest.cc]: Can we make the data for the unit test more minimal so that the test is not time consuming? Unit tests should be quite minimal, using very toy and simple data sets.

File: tests/unit/structure/GraphCut_unittest.cc [hunk 13]
```diff
@@ -168,411 +78,39 @@ TEST(GraphCut, graph_cut_chain)
...
+	CFGTestData* fg_test_data = new CFGTestData();
+	SG_REF(fg_test_data);
+	
+	EXPECT_NEAR(fg_test_data->test_sosvm(GRAPH_CUT), 0, 0.1);
 
-	SG_UNREF(labels_sgd);
-	SG_UNREF(sgd);
-	SG_UNREF(model);
-	SG_UNREF(fg_feats_train);
-	SG_UNREF(fg_labels_train);
```
Comment (iglesias) [inline tests/unit/structure/GraphCut_unittest.cc]: I get a bit scared with all these changes with `UNREF` and `REF`. Did you run valgrind and ensure nothing is leaking?

File: src/shogun/structure/FGTestData.h [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+
+#include <shogun/structure/MAPInference.h>
+#include <shogun/structure/StochasticSOSVM.h>
+
+#include <shogun/features/FactorGraphFeatures.h>
+#include <shogun/labels/FactorGraphLabels.h>
+
+namespace shogun
+{
+class CFGTestData: public CSGObject
```
Comment (iglesias) [inline src/shogun/structure/FGTestData.h]: Can you describe a little bit what is the function of this class?

File: src/shogun/structure/GEMPLP.h [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+
+#include <vector>
+#include <set>
+
+using namespace std;
+
+namespace shogun
+{
+#define IGNORE_IN_CLASSLIST
+/** GEMPLP (Generalized Max-product LP Relaxation) inference for fatcor graph
```
Comment (iglesias) [inline src/shogun/structure/GEMPLP.h]: Typo: factor graph.

File: src/shogun/structure/GEMPLP.h [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+#include <shogun/lib/SGNDArray.h>
+
+#include <vector>
+#include <set>
+
+using namespace std;
+
+namespace shogun
+{
+#define IGNORE_IN_CLASSLIST
```
Comment (iglesias) [inline src/shogun/structure/GEMPLP.h]: Why are you using this macro here?

Comment (iglesias) [inline src/shogun/structure/GEMPLP.h]: I am not saying is wrong though, just really asking :-)

File: src/shogun/structure/GEMPLP.h [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+
+#include <shogun/lib/config.h>
+#include <shogun/lib/SGVector.h>
+#include <shogun/structure/FactorGraph.h>
+#include <shogun/structure/Factor.h>
+#include <shogun/structure/MAPInference.h>
+#include <shogun/lib/SGNDArray.h>
+
+#include <vector>
+#include <set>
```
Comment (iglesias) [inline src/shogun/structure/GEMPLP.h]: Danger to have these includes if we want these headers to be public to modular interfaces!

File: src/shogun/structure/GEMPLP.h [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+ * Amir Globerson, Tommi Jaakkola
+ * Advances in Neural Information Processing Systems (NIPS). Vancouver, Canada. 2007.
+ *
+ * [2] Approximate Inference in Graphical Models using LP Relaxations.
+ * David Sontag
+ * Ph.D. thesis, Massachusetts Institute of Technology, 2010.
+ *
+ * The original implementation of GEMPLP can be found:
+ * http://cs.nyu.edu/~dsontag/code/mplp_ver2.tgz
+ * http://cs.nyu.edu/~dsontag/code/mplp_ver1.tgz
```
Comment (iglesias) [inline src/shogun/structure/GEMPLP.h]: All good with licenses here, I mean, is a new implementation what we have here inspired from the original?

**Comment by iglesias** [comment]
This looks like pretty amazing work, @Jiaolong. Great! Do you think you would be able to showcase in your structure output learning notebook what can we do with it?

File: tests/unit/structure/GEMPLP_unittest.cc (hunk 12 above)
Comment (Jiaolong) [inline tests/unit/structure/GEMPLP_unittest.cc]: Yes, thanks, I will try to make the data smaller.

File: tests/unit/structure/GraphCut_unittest.cc (hunk 13 above)
Comment (Jiaolong) [inline tests/unit/structure/GraphCut_unittest.cc]: I have run valgrind before, it seemed everything is fine. I can double check it.

File: src/shogun/structure/FGTestData.h (hunk 14 above)
Comment (Jiaolong) [inline src/shogun/structure/FGTestData.h]: Yes, I will add the descriptions.

File: src/shogun/structure/GEMPLP.h (hunk 16 above)
Comment (Jiaolong) [inline src/shogun/structure/GEMPLP.h]: Is this macro for avoiding generating python interface? Maybe I am wrong. :)

File: src/shogun/structure/GEMPLP.h (hunk 17 above)
Comment (Jiaolong) [inline src/shogun/structure/GEMPLP.h]: Yes, currently, it can not be used for modular interfaces. :(

File: src/shogun/structure/GEMPLP.h (hunk 18 above)
Comment (Jiaolong) [inline src/shogun/structure/GEMPLP.h]: Our implementation is not exactly the same as the original one, I mean it is not a copy & paste. However, we followed at least 80% of the original implementation for the core functions, e.g., `update_message()`. The main difference is that our version  is more compatible with the factor graph framework in shogun.

**Comment by Jiaolong** [comment]
@iglesias Thanks for your review! In fact, I have already prepared the notebook. An early version can be found here:
http://nbviewer.ipython.org/gist/Jiaolong/95633b7efd64bbd096f2

I will update it with more detailed explanations and probably with more experiments on other datasets.

File: tests/unit/structure/GraphCut_unittest.cc (hunk 13 above)
Comment (iglesias) [inline tests/unit/structure/GraphCut_unittest.cc]: If you already did it, then it should be fine ;-) I just wanted to be sure you did check it.

File: src/shogun/structure/GEMPLP.h (hunk 16 above)
Comment (iglesias) [inline src/shogun/structure/GEMPLP.h]: You don't need any macro to avoid swig from generating wrappers for a class. This is what you mean by generating Python interface. In fact, a class has to be included in swig files (.i files under interfaces directory) to be accessible to modular interfaces. We use macros when we make a class accessible to interfaces but we do not want some of its methods to be accessible via modular interfaces (but it is not this macro, see e.g. SGVector).

Comment (iglesias) [inline src/shogun/structure/GEMPLP.h]: Thus, I suggest to remove this line.

File: src/shogun/structure/GEMPLP.h (hunk 18 above)
Comment (iglesias) [inline src/shogun/structure/GEMPLP.h]: What does it mean to follow at least 80%? Is it that 80% of the lines are copy & paste but not 100% of them? :-)

File: src/shogun/structure/GEMPLP.h (hunk 17 above)
Comment (iglesias) [inline src/shogun/structure/GEMPLP.h]: I understand, however, that we can use gemplp from Python since you have used it in your notebook, right? I guess because you an enumerate or similar mechanism to select the type of approximate inference to use in your MAP class.

That is perfect, then! The interface, in this case, MAPInference class, is accessible to swig and through it one can define whether to use gemplp, etc.

File: src/shogun/structure/GEMPLP.h (hunk 16 above)
Comment (Jiaolong) [inline src/shogun/structure/GEMPLP.h]: OK, then I will remove it.

**Comment by iglesias** [comment]
I did not see mplp mentioned anywhere in the notebook. Maybe I overlooked?

File: src/shogun/structure/GEMPLP.h [hunk 19]
```diff
@@ -0,0 +1,152 @@
...
+ */
+IGNORE_IN_CLASSLIST class CGEMPLP: public CMAPInferImpl
```
Comment (iglesias) [inline src/shogun/structure/GEMPLP.h]: Remember to remove the ignore here as well.

File: src/shogun/structure/GEMPLP.h (hunk 17 above)
Comment (Jiaolong) [inline src/shogun/structure/GEMPLP.h]: Yes, you are right! This trick was done by Shell. :)

File: src/shogun/structure/GEMPLP.h (hunk 18 above)
Comment (Jiaolong) [inline src/shogun/structure/GEMPLP.h]: No, I mean the algorithm structure, the logic. :)

**Comment by Jiaolong** [comment]
Sorry, it was an old version. The only clue about MPLP is `infer_alg = LP_RELAXATION`. :p And now it should be `infer_alg = GEMPLP`.

File: src/shogun/structure/GEMPLP.h (hunk 19 above)
Comment (Jiaolong) [inline src/shogun/structure/GEMPLP.h]: I will do it.

File: src/shogun/structure/GEMPLP.h (hunk 18 above)
Comment (iglesias) [inline src/shogun/structure/GEMPLP.h]: We are all right, then.

**Comment by iglesias** [comment]
All right, got it.

Then, what we need for this pull request is:
1) Make unit tests simpler (probably just reducing input data) so we don't need to comment any of them, neither to wait long until they finish. If you want to have a larger test, then an integration one (that is, an example in Python modular) is the right place. Anyway, that would go to a future pull request.
2) Check the function of this FGTestData. I really wonder whether it makes sense to have it in C++. But I need to know more about it first.

Thanks a lot for getting back so quickly, @Jiaolong. Really cool.

**Comment by Jiaolong** [comment]
The reason I created `FGTestData` is that the same testing data are used in multiple unit tests, e.g., GrapCuts, GEMPLP, ect, and probably in the future new algroithms. Thus there are too many redudant code. Another reason is that, we haven't implemented functions to read factor graph from files, so we have to run these code to construct factor graph data.

**Comment by iglesias** [comment]
Then, it sounds like unit tests would be a better place for it, right?

**Comment by Jiaolong** [comment]
Yes, maybe we can move it into unit test folder. However, it can also be used in integral test.

File: src/shogun/structure/FGTestData.cpp [hunk 20]
```diff
@@ -0,0 +1,582 @@
...
+		v_feat.zero();
+
+		for (int32_t j = 0; j < 4 * (i + 1); j++)
+		{
+			int32_t r = random_indices[j];
+			v_feat[r] = 1;
+		}
+
+		for (int32_t f = 0; f < len_feat; f++)
+		{
```
Comment (iglesias) [inline src/shogun/structure/FGTestData.cpp]: One-liner loop, no need for braces. Here and below.

**Comment by iglesias** [comment]
I see your point. To use it from integration tests one could also just generate some data files and push them into the data repository. I admit this is not the most flexible solution though. I leave it up to you, then. If you want to keep the FG data class under structure, it is fine with me.

I will check for other data generators we have in Shogun and see whether it would make more sense there. I will let you know.

**Comment by iglesias** [comment]
There is this DataGenerator class inside the features directory. In my opinion, it is fine to keep this factor graph test data class here under structure, since it is quite specific. @karlnapf?

@Jiaolong, I do suggest to rename it to `FactorGraphDataGenerator`, to follow the convention with other data classes.

**Comment by Jiaolong** [comment]
All right, I will change the name.

**Comment by Jiaolong** [comment]
I just updated the code with your comments. Regarding to the unit tests, valgrind tests are fine:
```
shogun/build/tests/unit$ valgrind ./shogun-unit-test --gtest_filter="GEMP*"
==4564== Memcheck, a memory error detector
==4564== Copyright (C) 2002-2013, and GNU GPL'd, by Julian Seward et al.
==4564== Using Valgrind-3.10.0.SVN and LibVEX; rerun with -h for copyright info
==4564== Command: ./shogun-unit-test --gtest_filter=GEMP*
==4564== 
Note: Google Test filter = GEMP*
[==========] Running 7 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 7 tests from GEMPLP
[ RUN      ] GEMPLP.find_intersections_index
[       OK ] GEMPLP.find_intersections_index (27 ms)
[ RUN      ] GEMPLP.max_in_subdimension
[       OK ] GEMPLP.max_in_subdimension (8 ms)
[ RUN      ] GEMPLP.initialization
[       OK ] GEMPLP.initialization (64 ms)
[ RUN      ] GEMPLP.convert_energy_to_potential
[       OK ] GEMPLP.convert_energy_to_potential (6 ms)
[ RUN      ] GEMPLP.simple_chain
[       OK ] GEMPLP.simple_chain (32 ms)
[ RUN      ] GEMPLP.random_chain
[       OK ] GEMPLP.random_chain (34 ms)
[ RUN      ] GEMPLP.sosvm
[       OK ] GEMPLP.sosvm (30277 ms)
[----------] 7 tests from GEMPLP (30458 ms total)

[----------] Global test environment tear-down
[==========] 7 tests from 1 test case ran. (30498 ms total)
[  PASSED  ] 7 tests.
==4564== 
==4564== HEAP SUMMARY:
==4564==     in use at exit: 0 bytes in 0 blocks
==4564==   total heap usage: 7,657,410 allocs, 7,657,410 frees, 164,492,480 bytes allocated
==4564== 
==4564== All heap blocks were freed -- no leaks are possible
==4564== 
==4564== For counts of detected and suppressed errors, rerun with: -v
==4564== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

File: src/shogun/structure/FactorGraphDataGenerator.h [hunk 21]
```diff
@@ -0,0 +1,138 @@
...
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Written (W) 2015 Jiaolong Xu
+ * Copyright (C) 2015 Jiaolong Xu
+ */
+
+#ifndef __FGDATA_H__
```
Comment (iglesias) [inline src/shogun/structure/FactorGraphDataGenerator.h]: Update the name here as well, please, and in the line below this one.

**Comment by iglesias** [comment]
Great, thanks a lot for the valgrind log. I see however the GEMPLP.sosvm is still taking too long. Can you reduce that time, please?

**Comment by Jiaolong** [comment]
The time shown above is measured under valgrind, so it is longer than running at realse mode. It is `452 ms` on my PC. But anyway, I will try to reduce more. Currently, the SOSVM data has 4 classes and 16 samples.

File: src/shogun/structure/FactorGraphDataGenerator.h (hunk 21 above)
Comment (Jiaolong) [inline src/shogun/structure/FactorGraphDataGenerator.h]: Oops!

**Comment by Jiaolong** [comment]
Now the unit test time is reduced to 160 ms.
 `[       OK ] GEMPLP.sosvm (160 ms)`

**Comment by iglesias** [comment]
Oh, sure. Sorry, I did not take into account valgrind when I looked at the running time. 0.4 seconds should be fine, no need to worry more about it.

**Comment by iglesias** [comment]
Perfect, we are ready to go, then. Thanks, @Jiaolong.

**Comment by Jiaolong** [comment]
Many thanks! :D

**Comment by iglesias** [comment]
Feel free to send an update for the notebook when you have the time. It would be great if it includes as well some short discussion about gemplp.

**Comment by Jiaolong** [comment]
Sure! I will try to polish the notebook and send it shortly. Thanks!

**Comment by hushell** [comment]
@Jiaolong @iglesias It's a great news this patch is merged! Sorry I couldn't help for the merge. I spent last fews day on traveling. Thanks for your hardworking! Let's plan our next steps :)
